### PR TITLE
Support written decades as range in English

### PR DESCRIPTION
**Pull request by lucinezhang**
github issue #211

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Definitions/French/DateTimeDefinitions.cs [hunk 1]
```diff
@@ -631,5 +634,9 @@ public static class DateTimeDefinitions
 			{ "newyeareve", new string[] { "réveillondenouvelan", "reveillondenouvelan", "lasaint-sylvestre", "lasaintsylvestre" } }
 		};
 		public const string NightRegex = @"\b(minuit|nuit)\b";
+		public static readonly Dictionary<string, int> WrittenDecades = new Dictionary<string, int>
+		{
+			{ "xxx", 20 }
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions/French/DateTimeDefinitions.cs]: The dictionaries for the languages not yet supported could have been left empty, no?

File: .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs [hunk 2]
```diff
@@ -197,6 +197,9 @@ public static class DateTimeDefinitions
 		public static readonly string NumberAsTimeRegex = $@"\b({EngTimeRegex}|{PeriodHourNumRegex}|{BaseDateTime.HourRegex})\b";
 		public static readonly string TimeBeforeAfterRegex = $@"\b(((?<=\b(before|after)\s+)({EngTimeRegex}|{HourNumRegex}|{BaseDateTime.HourRegex}|{MidTimeRegex}))|{MidTimeRegex})\b";
 		public const string DateNumberConnectorRegex = @"^\s*(?<connector>\s+at)\s*$";
+		public const string CenturyRegex = @"(?<century>one|two|three|four|five|six|seven|eight|nine|ten|eleven|twelve|thirteen|fourteen|fifteen|sixteen|seventeen|eighteen|nineteen|twenty|twenty one|twenty two)";
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: Why is this regex needed?

File: .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs [hunk 3]
```diff
@@ -541,5 +544,16 @@ public static class DateTimeDefinitions
...
+		public static readonly Dictionary<string, int> WrittenDecades = new Dictionary<string, int>
+		{
+			{ "twenties", 20 },
+			{ "thirties", 30 },
+			{ "forties", 40 },
+			{ "fifties", 50 },
+			{ "sixties", 60 },
+			{ "seventies", 70 },
+			{ "eighties", 80 },
+			{ "nineties", 90 }
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: In UK English the decade of 2000-09 is usually know as "the noughties". This should resolve to that range.

File: Specs/DateTime/English/DatePeriodParser.json [hunk 4]
```diff
@@ -2337,4 +2337,160 @@
...
+          "startDate": "1970-01-01",
+          "endDate": "1979-12-31"
+        }
+      }
+    }
+  ]
+}
+,
+{
+  "Input": "In the 70's",
```
Comment (tellarin) [inline Specs/DateTime/English/DatePeriodParser.json]: We should also support "'40s".

File: Specs/DateTime/English/DatePeriodExtractor.json [hunk 5]
```diff
@@ -2649,4 +2649,76 @@
     }
   ]
 }
+,
+{
+  "Input": "In the 1970s",
```
Comment (tellarin) [inline Specs/DateTime/English/DatePeriodExtractor.json]: Does "the 2000s" work too? Can we add a test case for it?

File: .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs (hunk 3 above)
Comment (tellarin) [reply .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: Somewhat similarly, "the twenty-tens" should resolve to 2010-2019.

File: Specs/DateTime/English/DatePeriodParser.json [hunk 6]
```diff
@@ -2337,4 +2337,160 @@
...
+          "startDate": "1970-01-01",
+          "endDate": "1979-12-31"
+        }
+      }
+    }
+  ]
+}
+,
+{
+  "Input": "In the nineteen seventies",
```
Comment (tellarin) [inline Specs/DateTime/English/DatePeriodParser.json]: What about "the two thousand and tens"?

File: .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs (hunk 3 above)
Comment (tellarin) [reply .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: "the two thousands" could also be tagged.

Comment (tellarin) [reply .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: In many of these cases for the written form the "the" has to be there.

File: .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs (hunk 2 above)
Comment (lucinezhang) [reply .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: I need this to specify the written century.

File: .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs (hunk 3 above)
Comment (lucinezhang) [reply .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: Added.

File: .NET/Microsoft.Recognizers.Definitions/French/DateTimeDefinitions.cs (hunk 1 above)
Comment (lucinezhang) [reply .NET/Microsoft.Recognizers.Definitions/French/DateTimeDefinitions.cs]: I didn't find a way to left the dictionary empty in yaml.

File: Specs/DateTime/English/DatePeriodExtractor.json (hunk 5 above)
Comment (lucinezhang) [reply Specs/DateTime/English/DatePeriodExtractor.json]: Add the test case. It works.

File: Specs/DateTime/English/DatePeriodParser.json (hunk 4 above)
Comment (lucinezhang) [reply Specs/DateTime/English/DatePeriodParser.json]: It's already supported. There's a test case.

File: Specs/DateTime/English/DatePeriodParser.json (hunk 6 above)
Comment (lucinezhang) [reply Specs/DateTime/English/DatePeriodParser.json]: Support now. Added a test case.

File: .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs [hunk 7]
```diff
@@ -541,5 +544,19 @@ public static class DateTimeDefinitions
 			{ "christmaseve", new string[] { "christmaseve" } },
 			{ "newyeareve", new string[] { "newyearseve", "newyeareve" } }
 		};
+		public static readonly Dictionary<string, int> WrittenDecades = new Dictionary<string, int>
+		{
+			{ "noughties", 0 },
+			{ "tens", 10 },
+			{ "twenties", 20 },
+			{ "two thousands", 20 },
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: "two thousands" is 2000s, not 20s. This needs fixing.
Same for "noughties"
"twenty-tens" should also only covers 2010s.  Are you sure these cases cover only the right occurrences? I can't see that from the tests.

Comment (lucinezhang) [reply .NET/Microsoft.Recognizers.Definitions/English/DateTimeDefinitions.cs]: Fixed and added some test cases.

**Pull request merged**